src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerSessionMonitor.cs

```diff
@@ -50,35 +72,111 @@ public void AddSessionListener(IProfilerSessionListener listener)
         /// <summary>
         /// Start monitoring the provided sessions
         /// </summary>
-        public bool StartMonitoringSession(ProfilerSession session)
+        public bool StartMonitoringSession(string viewerID, IXEventSession session)
         {
             lock (this.sessionsLock)
             {
-                // start the monitoring thread 
+                // start the monitoring thread
                 if (this.processorThread == null)
                 {
-                    this.processorThread = Task.Factory.StartNew(ProcessSessions);;
+                    this.processorThread = Task.Factory.StartNew(ProcessSessions);
+                }
+
+                // create new profiling session if needed
+                if (!this.monitoredSessions.ContainsKey(session.ID))
+                {
+                    var profilerSession = new ProfilerSession();
+                    profilerSession.XEventSession = session;
+
+                    this.monitoredSessions.Add(session.ID, profilerSession);
                 }
 
-                if (!this.monitoredSessions.ContainsKey(session.SessionId))
+                // create a new viewer, or configure existing viewer
+                Viewer viewer;
+                if(!this.allViewers.TryGetValue(viewerID, out viewer))
                 {
-                    this.monitoredSessions.Add(session.SessionId, session);
+                    viewer = new Viewer(viewerID, true, session.ID);
+                    allViewers.Add(viewerID, viewer);
+                }
+                else
+                {
+                    viewer.active = true;
+                    viewer.xeSessionID = session.ID;
+                }
+
+                // add viewer to XEvent session viewers
+                List<string> viewers;
+                if(this.sessionViewers.TryGetValue(session.ID, out viewers))
+                {
+                    viewers.Add(viewerID);
+                }
+                else
+                {
+                    viewers = new List<string>{ viewerID };
+                    sessionViewers.Add(session.ID, viewers);
                 }
             }
 
             return true;
         }
 
+        // TODO: Clean this up
+        /// <summary>
+        /// Stop monitoring the session watched by viewerID
+        /// </summary>
+        public bool StopMonitoringSession(string viewerID, out ProfilerSession session)
+        {
+            lock (this.sessionsLock)
```

Comment:
I don't see where this code actually stops the session on the server side. Does that happen somewhere else or is it missing?

Reply:
This code is purely for stopping the monitoring of the session, because session monitor only manages what's being polled. it's called by the profiler service's HandleStopProfilingRequest method, and that's where the actual session is stopped.